source/agora/common/Block.d

```diff
@@ -190,32 +197,42 @@ public struct Block
 
     ***************************************************************************/
 
-    public Hash buildMerkleTree() nothrow @safe
+    public Hash buildMerkleTree () nothrow @safe
     {
-        this.merkle_tree.reserve(this.txs.length);
-        foreach (ref tx; this.txs)
-            this.merkle_tree ~= hashFull(tx);
+        this.merkle_tree.length = (this.txs.length * 2) - 1;
+        return this.buildMerkleTreeImpl();
+    }
+
+    /// Ditto
+    public Hash buildMerkleTreeImpl () nothrow @safe @nogc
+    {
+        import core.bitop;
+        const log2 = bsf(this.txs.length);
+        assert((1 << log2) == this.txs.length,
+               "Transactions for a block should be a strict power of 2");
+        assert(this.merkle_tree.length == (this.txs.length * 2) - 1);
+
+        foreach (size_t idx, ref hash; this.merkle_tree[0 .. this.txs.length])
+            hash = hashFull(this.txs[idx]);
 
         // transactions are ordered lexicographically by hash in the Merkle tree
-        this.merkle_tree.sort!("a < b");
+        this.merkle_tree[0 .. this.txs.length].sort!("a < b");
 
-        size_t j = 0;
-        for (size_t length = this.txs.length; length > 1; length = (length + 1) / 2)
+        immutable len = this.merkle_tree.length;
+        for (size_t order = 0; order < log2; order++)
         {
-            for (size_t i = 0; i < length; i += 2)
-            {
-                size_t i2 = min(i + 1, length - 1);
-                this.merkle_tree ~= mergeHash(this.merkle_tree[j + i],
-                    this.merkle_tree[j + i2]);
-            }
-
-            j += length;
+            immutable start = len - (len >> (order));
+            immutable end   = len - (len >> (order + 1));
+            this.merkle_tree[start .. end].chunks(2)
+                .map!(tup => mergeHash(tup[0], tup[1]))
+                .enumerate(size_t(end))
+                .each!(val => this.merkle_tree[val[0]] = val[1]);
```

Comment:
If you have enumerate there, I believe you can re-write this as:

```
.each!((idx, val) => this.markle_tree[idx] = val);
```

Reply:
Thanks, fixed